provider/pkg/provider/container.go

```diff
@@ -698,3 +702,152 @@ func updateNodePoolConfig(diffPropPath, targetPropPath resource.PropertyPath,
+}
+
+func (*providerResourceState) ElementType() reflect.Type {
+	return reflect.TypeOf((*callableResource)(nil)).Elem()
+}
+
+type customResourceState struct {
+	pulumi.CustomResourceState
+	name    string
+	Outputs pulumi.MapOutput `pulumi:""`
+}
+
+// GetOutput returns the named output of the resource.
+func (st *customResourceState) GetOutput(k string) pulumi.Output {
+	return st.Outputs.ApplyT(func(outputs map[string]interface{}) (interface{}, error) {
+		out, ok := outputs[k]
+		if !ok {
+			return nil, fmt.Errorf("no output '%s' on resource '%s'", k, st.name)
+		}
+		return out, nil
+	})
+}
+
+func (st *customResourceState) CustomResource() *pulumi.CustomResourceState {
+	return &st.CustomResourceState
+}
+
+func (st *customResourceState) ProviderResource() *pulumi.ProviderResourceState {
+	return nil
+}
+
+func (*customResourceState) ElementType() reflect.Type {
+	return reflect.TypeOf((*callableResource)(nil)).Elem()
+}
+
+// clusterGetKubeconfigResult is the representation of the result from the getKubeconfig method call.
+// Note the use of `liftSingleValueMethodReturns` in the schema is a codegen construct and applies
+// in the consuming SDK. The schema still only supports object types in the output.
+type clusterGetKubeconfigResult struct {
+	Kubeconfig pulumi.StringOutput `pulumi:"kubeconfig"`
+}
+
+const kubeconfigFmt = `apiVersion: v1
+clusters:
+- cluster:
+    certificate-authority-data: %[3]s
+    server: https://%[2]s
+  name: %[1]s
+contexts:
+- context:
+    cluster: %[1]s
+    user: %[1]s
+  name: %[1]s
+current-context: %[1]s
+kind: Config
+preferences: {}
+users:
+- name: %[1]s
+  user:
+    exec:
+      apiVersion: client.authentication.k8s.io/v1beta1
+      command: gke-gcloud-auth-plugin
+      installHint: Install gke-gcloud-auth-plugin for use with kubectl by following
+        https://cloud.google.com/blog/products/containers-kubernetes/kubectl-auth-changes-in-gke
+      provideClusterInfo: true
+`
+
+// getKubeConfigCallHandler returns a function which converts the getKubeconfig Call GRPC invocation to Pulumi's Go
+// programming model.
+func getKubeConfigCallHandler(label, tok string, callArgs resource.PropertyMap) func(*pulumi.Context, string,
+	pulumiprov.CallArgs) (*pulumiprov.CallResult, error) {
+	return func(ctx *pulumi.Context, tok string,
+		args pulumiprov.CallArgs) (*pulumiprov.CallResult, error) {
+		var res pulumi.Resource
+		var state callableResource
+		// If we get here, callArgs will already have a `__self__` and will be known resource reference.
+		self := callArgs["__self__"]
+		if !self.IsResourceReference() {
+			return nil, fmt.Errorf("call(%s) __self__ was expected to be a resource reference", tok)
+		}
+		ref := self.ResourceReferenceValue()
+		if strings.HasPrefix(string(ref.URN.Type()), "pulumi:providers:") {
+			r := providerResourceState{name: ""}
+			res = &r
+			state = &r
+		} else {
+			r := customResourceState{name: ""}
+			res = &r
+			state = &r
+		}
```

Comment:
I think this check and the associated `providerResourceState` struct can be deleted. You're only defining the methods on the two versions of the `Cluster` custom resource, so you're not going to be getting a resource reference to a provider resource in `__self__` here, only the `Cluster` custom resources.

Reply:
> I think this check and the associated `providerResourceState` struct can be deleted. You're only defining the methods on the two versions of the `Cluster` custom resource, so you're not going to be getting a resource reference to a provider resource in `__self__` here, only the `Cluster` custom resources.

Sweet - thanks for the clarification! Removed.